syntax = "proto3";

package arl.sidecar;

option go_package = "github.com/Lincyaw/agent-env/pkg/pb";

// AgentService provides command execution and interactive shell
service AgentService {
  // Execute runs a command (Job mode) or starts a background service (Service mode)
  rpc Execute(ExecRequest) returns (stream ExecLog);

  // InteractiveShell provides bidirectional streaming for interactive shell sessions
  rpc InteractiveShell(stream ShellInput) returns (stream ShellOutput);
}

// ExecRequest specifies a command to execute
message ExecRequest {
  repeated string command = 1;
  map<string, string> env = 2;
  string working_dir = 3;
  bool background = 4;  // true for Service mode, false for Job mode
  int32 timeout_seconds = 5;
}

// ExecLog streams output from command execution
message ExecLog {
  string stdout = 1;
  string stderr = 2;
  int32 exit_code = 3;
  bool done = 4;
}

// ShellInput represents input to an interactive shell session
message ShellInput {
  oneof input {
    string data = 1;      // Stdin data to send to the shell
    string signal = 2;    // Signal to send (e.g., "SIGINT" for Ctrl+C)
  }
  bool resize = 3;        // Terminal resize event
  int32 rows = 4;         // Terminal rows (used with resize)
  int32 cols = 5;         // Terminal columns (used with resize)
}

// ShellOutput represents output from an interactive shell session
message ShellOutput {
  string data = 1;        // Stdout/stderr data from the shell
  int32 exit_code = 2;    // Exit code (set when shell exits)
  bool closed = 3;        // True when shell session has ended
}
